gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManagementTaskStreamImpl.java

```diff
@@ -0,0 +1,159 @@
+  private final Config config;
+  @Getter private final EventSubmitter eventSubmitter;
+  @Getter private static final DagManagerMetrics dagManagerMetrics = new DagManagerMetrics();
+  private volatile boolean isActive = false;
+
+  @Inject(optional=true)
+  protected Optional<DagActionStore> dagActionStore;
+  @Inject
+  @Getter DagManagementStateStore dagManagementStateStore;
+  private static final int MAX_HOUSEKEEPING_THREAD_DELAY = 180;
+  private final BlockingQueue<DagActionStore.DagAction> dagActionQueue = new LinkedBlockingQueue<>();
+
+  @Inject
+  public DagManagementTaskStreamImpl(Config config, Optional<DagActionStore> dagActionStore, DagManagementStateStore dagManagementStateStore) {
+    this.config = config;
+    this.dagActionStore = dagActionStore;
+    this.dagManagementStateStore = dagManagementStateStore;
+    MetricContext metricContext = Instrumented.getMetricContext(ConfigUtils.configToState(ConfigFactory.empty()), getClass());
+    this.eventSubmitter = new EventSubmitter.Builder(metricContext, "org.apache.gobblin.service").build();
+  }
+
+  public void setActive(boolean active) {
+    if (this.isActive == active) {
+      log.info("DagManagementTaskStreamImpl already {}, skipping further actions.", (!active) ? "inactive" : "active");
+    }
+    this.isActive = active;
+    try {
+      if (this.isActive) {
+        log.info("Activating DagManagementTaskStreamImpl.");
+        //Initializing state store for persisting Dags.
+        this.dagManagementStateStore.start();
+        dagManagerMetrics.activate();
+      } else { //Mark the DagManager inactive.
+        log.info("Inactivating the DagManagementTaskStreamImpl. Shutting down all DagManager threads");
+        dagManagerMetrics.cleanup();
+      }
+    } catch (IOException e) {
+        log.error("Exception encountered when activating the DagManagementTaskStreamImpl", e);
+        throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public synchronized void addDagAction(DagActionStore.DagAction dagAction) throws IOException {
+    // TODO: Used to track missing dag issue, remove later as needed
+    log.info("Add dagAction{}", dagAction);
+    if (!isActive) {
+      log.warn("Skipping add dagAction because this instance of DagManagementTaskStreamImpl is not active for dag: {}",
+          dagAction);
+      return;
+    }
+
+    DagManager.DagId dagId = DagManagerUtils.generateDagId(dagAction.getFlowGroup(), dagAction.getFlowName(), dagAction.getFlowExecutionId());
+    String dagIdString = dagId.toString();
+    if (this.dagManagementStateStore.containsDag(dagId)) {
+      log.warn("Already tracking a dag with dagId {}, skipping.", dagIdString);
+      return;
+    }
+
+    // After persisting the dag, its status will be tracked by active dagManagers so the action should be deleted
+    // to avoid duplicate executions upon leadership change
+    if (this.dagActionStore.isPresent()) {
+      this.dagActionStore.get().deleteDagAction(dagAction);
+    }
```

Comment:
seems naively that if multiple active participants that they'd all try to delete the same dag action from the shared store.

or are you relying on some trick, such as that `dagActionStore.isPresent()` is only true for active participants